File: src/validation/validator.py
"""
Validation: check ExtractionCandidate against Rule (empty, max_length, required).
"""

from src.rules.models import ExtractionCandidate, Rule


def validate(candidate: ExtractionCandidate, rule: Rule) -> tuple[bool, str]:
    """
    Validate candidate against rule. Checks: empty, max_length if set, required logic.
    Returns (is_valid, error_message). error_message is "" when valid.
    """
    value = candidate.normalized_value

    if rule.required and (value is None or value == ""):
        return False, "required field is empty"

    if rule.max_length is not None and len(value) > rule.max_length:
        return False, f"value length {len(value)} exceeds max_length {rule.max_length}"

    return True, ""

File: src/validation/validator.cpp
#include "document_ingestion/validator.hpp"

namespace document_ingestion {

std::tuple<bool, std::string> validate(const ExtractionCandidate& candidate, const Rule& rule) {
  if (rule.required && candidate.raw_match.empty()) {
    return {false, "required rule produced no match"};
  }
  if (rule.max_length && candidate.normalized_value.size() > static_cast<size_t>(*rule.max_length)) {
    return {false, "value exceeds max_length"};
  }
  return {true, ""};
}

}  // namespace document_ingestion
